orc8r/cloud/go/services/analytics/calculations/user_state.go

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright 2020 The Magma Authors.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package calculations
+
+import (
+	"magma/orc8r/cloud/go/services/analytics/query_api"
+	"magma/orc8r/lib/go/metrics"
+
+	"github.com/golang/glog"
+)
+
+//UserNetworkState contains the user state per network
+type UserNetworkState struct {
+	TotalUsersPerNetwork int
+	UsersGatewayTable    map[string]int
+}
+
+// UserStateMgr contains the user state in the deployment to enforce minUserThreshold constraint
+type UserStateMgr struct {
```

Comment:
Prefer interfaces instead of public structs. Make `UserStateManager` an interface, with clear indications of the interface contract (including clear doc comments). Then make a `userStateManager` private struct that implements (with all-private fields), and a `NewUserStateManager` whose public signature returns the interface, but its implementation returns the private struct

See the `remoteIndexer` for inspiration services/state/indexer/remote_indexer.go:42

Reply:
I can understand abstracting it out when this is going to be implemented in a couple of ways. If there is going to be only one UserStateManager, what is the strong reason for making this an interface ?